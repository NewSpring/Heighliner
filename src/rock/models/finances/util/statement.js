/*
 *
 * this is a temporary custom render for statements
 * after we land passing variables to the RenderLavaTemplate api
 * we can deprecated this and move to looking up a merge template and rendering
 * it instead.
*/

import React from "react";
import ReactDOMServer from "react-dom/server";
import pdf from "html-pdf";
import uuid from "node-uuid";

export const generatePDF = (component) => {
  const html = ReactDOMServer.renderToStaticMarkup(component);
  console.log("component = ", component);
  console.log("html = ", html);
  return new Promise((r, f) => {
    // XXX James says this isn't really worth mocking independent parts
    // of pdf.create. So instead we just verify it fails, or returns a base64 stringify
    // from a given react component
    pdf
      .create(html, {
        format: "letter",
        border: {
          top: "0.6in",
          right: "0.6in",
          bottom: "0.6in",
          left: "0.6in",
        },
      })
      .toBuffer((err, buffer) => {
        if (err) return f(err);

        r(buffer.toString("base64"));
      });
  });
};

export const formatMoney = amount => `$${amount.toFixed(2).replace(/\B(?=(\d{3})+(?!\d))/g, ",")}`;

export const Statement = ({ transactions, person, home, total }) => (
  <html>
    <head>
      <style>{`
        html,
        body {
          height: 100%;
        }
        @media all {
          .page-break {
            display: block;

            page-break-before: always;
          }
        }

        @media print {
          .page-break {
            display: block;

            page-break-before: always;
          }
        }
        .container {
          margin: 0 auto;
        }

        .pre-text {
          font-family: "Courier New";
          font-size: 12px;
          line-height: 1.4em;
        }

        .print-subheader {
          color: black;
          font-family: "Helvetica";
        }

        #logo-img {
          margin-top: 1000px
          width: 155px;
          height: 40px;
        }

        #church-address {
          float: left;
          margin-top: 0;
          padding-right: 150px;
        }

        #verse {
          display: inline-block;
        }

        #statement-table td {
          padding-top: 10px;
          padding-bottom: 10px;

          border-bottom: 1px solid #ddd;
        }

        th {
          padding-bottom: 10px;

          border-bottom: 1px solid #ddd;

          font-family: "Helvetica";
          font-weight: 700;
        }

        table {
          text-align: left;

          font-family: "Courier New";
          font-size: 12px;
        }

        h7 {
          font-size: 12px;
        }

        .type {
          width: 55%;
        }

        .total {
          color: black;

          font-family: "Helvetica";
          font-size: 18px;
          font-weight: 800;
        }

        .newspring--icon,
        .newspring--icon::after {
          background-color: black;
        }

        .newspring--icon {
          position: absolute;

          margin-top: 200px;
          margin-left: 200px;
        }

        .logo-icon {
          position: absolute;
          right: 0;
          bottom: 0;

          margin: 20px;
        }

        #statement-table {
          padding-top: 20px;
        }

        #statement-table .total td {
          padding-top: 20px;
          padding-bottom: 20px;
        }

        .footer-text {

          width: 6in;
          margin-top: 40px;
          margin-bottom: 100px;

          font-size: 12px;
          font-style: italic;
          line-height: 1.4em;
        }

      `}</style>
    </head>
    <body>
      <div className="container">
        <p>
          <br />
          <br />
          <br />
        </p>
        <img
          id="logo-img"
          src="https://s3.amazonaws.com/ns.images/newspring/icons/newspring-church-logo-black.png"
        />
        <h2
          style={{ fontFamily: "Helvetica" }}
          className="print-subheader soft-sides soft-half-bottom"
        >
          Contribution Statement
        </h2>

        <div className="soft pre-text">
          <p id="church-address">
            NewSpring Church
            <br />
            EIN# 26-4189337 <br />
            PO Box 1407
            <br />
            Anderson, SC 29622 US
          </p>

          <p>
            {person.NickName || person.FirstName} {person.LastName}
            <br />
            {home.Street1}
            <br />
            {home.Street2 && (
              <span>
                {home.Street2}
                <br />
              </span>
            )}
            {home.City} {home.State} {home.PostalCode}
          </p>

          <p id="verse">
            Malachi 3:10 "Bring the whole tithe into the storehouse, that there may be food in my
            house. Test me in this,' says the LORD Almighty, 'and see if I will not throw open the
            floodgates of heaven and pour out so much blessing that you will not have room enough
            for it."
          </p>
        </div>

        <table className="soft" style={{ width: "100%" }} id="statement-table">
          <tbody>
            <tr>
              <th className="type" style={{ fontFamily: "Helvetica", fontWeight: 700 }}>
                Account
              </th>
              <th style={{ fontFamily: "Helvetica", fontWeight: 700 }}>Date</th>
              <th style={{ fontFamily: "Helvetica", fontWeight: 700 }}>Amount</th>
            </tr>
            {transactions.map((transaction, key) => (
              <tr key={key}>
                <td>{transaction.Name}</td>
                <td>{transaction.Date}</td>
                <td>{formatMoney(transaction.Amount)}</td>
              </tr>
            ))}
            <tr className="total" style={{ width: "100%" }}>
              <td className="soft-ends" style={{ fontFamily: "Helvetica" }}>
                Total
              </td>
              <td />
              <td className="soft-ends total-amount" style={{ fontFamily: "Helvetica" }}>
                {formatMoney(total)}
              </td>
            </tr>
          </tbody>
        </table>

        <p className="footer-text soft">
          Thank you for your continued support. Please note that for taxation purposes, no goods or
          services were received in return for these contributions. If you have any questions or
          concerns, please contact us at giving@newspring.cc or 864-965-9990.
        </p>
      </div>
    </body>
  </html>
);

export default props => generatePDF(<Statement {...props} />);
