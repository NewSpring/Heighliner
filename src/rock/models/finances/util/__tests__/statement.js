import React from "react";
import renderer from "react-test-renderer";
import { formatMoney, Statement, generatePDF } from "../statement";
import ReactDOMServer from "react-dom/server";

describe("formatMoney", () => {
  it("should render 2 decimal places", () => {
    expect(formatMoney(1)).toEqual("$1.00");
  });

  it("should format with commas", () => {
    expect(formatMoney(1234)).toEqual("$1,234.00");
    expect(formatMoney(1234566)).toEqual("$1,234,566.00");
  });

  it("should allow non-whole numbers", () => {
    expect(formatMoney(1234.6)).toEqual("$1,234.60");
    expect(formatMoney(0.65)).toEqual("$0.65");
  });
});

describe("Statement", () => {
  const person = { FirstName: "", LastName: "", nickName: "" };
  const home = {
    Street1: "3400 Vine St",
    Street2: "",
    City: "Cincinnati",
    State: "OH",
    PostalCode: "45220",
  };
  const transactions = [
    { Name: "Admission", Date: "2016-01-01", Amount: 15 },
    { Name: "Admission", Date: "2016-02-01", Amount: 15 },
  ];
  const total = 30;

  it("should render with data", () => {
    const component = renderer.create(
      <Statement
        person={person}
        home={home}
        transactions={transactions}
        total={total}
      />,
    );
    expect(component).toMatchSnapshot();
  });
});

describe("generatePDF", () => {
  it("should call pdf.create with markup and settings", async () => {
    jest.mock("react-dom/server");
    ReactDOMServer.renderToStaticMarkup = jest.fn(c => c);
    const results = await generatePDF("<div>Hello</div>");

    // XXX can't properly snapshot the results, because of a creationDate
    // being injected by pdf.create (which we don't have time to properly mock)
    // can use Buffer(results, "base64").toString() to print
    expect(results).toBeDefined();
  });

  it("should fail with no component to render", async () => {
    jest.mock("react-dom/server");
    ReactDOMServer.renderToStaticMarkup = jest.fn(c => c);
    return generatePDF().then(
      res => {
        throw new Error("generatePDF didn't fail");
      },
      res => {
        expect(res).toBeDefined();
      },
    );
  });
});
