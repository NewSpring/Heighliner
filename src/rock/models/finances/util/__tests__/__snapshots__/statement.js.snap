exports[`Statement should render with data 1`] = `
<html>
  <head>
    <style>
      
              html,
              body {
                height: 100%;
              }
              @media all {
                .page-break {
                  display: block;
      
                  page-break-before: always;
                }
              }
      
              @media print {
                .page-break {
                  display: block;
      
                  page-break-before: always;
                }
              }
              .container {
                margin: 0 auto;
              }
      
              .pre-text {
                font-family: \"Courier New\";
                font-size: 12px;
                line-height: 1.4em;
              }
      
              .print-subheader {
                color: black;
                font-family: \"Helvetica\";
              }
      
              #logo-img {
                margin-top: 1000px
                width: 155px;
                height: 40px;
              }
      
              #church-address {
                float: left;
                margin-top: 0;
                padding-right: 150px;
              }
      
              #verse {
                display: inline-block;
              }
      
              #statement-table td {
                padding-top: 10px;
                padding-bottom: 10px;
      
                border-bottom: 1px solid #ddd;
              }
      
              th {
                padding-bottom: 10px;
      
                border-bottom: 1px solid #ddd;
      
                font-family: \"Helvetica\";
                font-weight: 700;
              }
      
              table {
                text-align: left;
      
                font-family: \"Courier New\";
                font-size: 12px;
              }
      
              h7 {
                font-size: 12px;
              }
      
              .type {
                width: 55%;
              }
      
              .total {
                color: black;
      
                font-family: \"Helvetica\";
                font-size: 18px;
                font-weight: 800;
              }
      
              .newspring--icon,
              .newspring--icon::after {
                background-color: black;
              }
      
              .newspring--icon {
                position: absolute;
      
                margin-top: 200px;
                margin-left: 200px;
              }
      
              .logo-icon {
                position: absolute;
                right: 0;
                bottom: 0;
      
                margin: 20px;
              }
      
              #statement-table {
                padding-top: 20px;
              }
      
              #statement-table .total td {
                padding-top: 20px;
                padding-bottom: 20px;
              }
      
              .footer-text {
      
                width: 6in;
                margin-top: 40px;
                margin-bottom: 100px;
      
                font-size: 12px;
                font-style: italic;
                line-height: 1.4em;
              }
      
            
    </style>
  </head>
  <body>
    <div
      className="container">
      <p>
        <br />
        <br />
        <br />
      </p>
      <img
        id="logo-img"
        src="https://s3.amazonaws.com/ns.images/newspring/icons/newspring-church-logo-black.png" />
      <h2
        className="print-subheader soft-sides soft-half-bottom"
        style={
          Object {
            "fontFamily": "Helvetica",
          }
        }>
        Contribution Statement
      </h2>
      <div
        className="soft pre-text">
        <p
          id="church-address">
          NewSpring Church  
          <br />
          EIN# 26-4189337 
          <br />
          PO Box 1407  
          <br />
          Anderson, SC 29622 US
        </p>
        <p>
          
           
          
          <br />
          3400 Vine St
          <br />
          
          Cincinnati
           
          OH
           
          45220
        </p>
        <p
          id="verse">
          Malachi 3:10 \"Bring the whole tithe into the storehouse, that there may be food in my house. Test me in this,\' says the LORD Almighty, \'and see if I will not throw open the floodgates of heaven and pour out so much blessing that you will not have room enough for it.\"
        </p>
      </div>
      <table
        className="soft"
        id="statement-table"
        style={
          Object {
            "width": "100%",
          }
        }>
        <tbody>
          <tr>
            <th
              className="type"
              style={
                Object {
                  "fontFamily": "Helvetica",
                  "fontWeight": 700,
                }
              }>
              Account
            </th>
            <th
              style={
                Object {
                  "fontFamily": "Helvetica",
                  "fontWeight": 700,
                }
              }>
              Date
            </th>
            <th
              style={
                Object {
                  "fontFamily": "Helvetica",
                  "fontWeight": 700,
                }
              }>
              Amount
            </th>
          </tr>
          <tr>
            <td>
              Admission
            </td>
            <td>
              2016-01-01
            </td>
            <td>
              $15.00
            </td>
          </tr>
          <tr>
            <td>
              Admission
            </td>
            <td>
              2016-02-01
            </td>
            <td>
              $15.00
            </td>
          </tr>
          <tr
            className="total"
            style={
              Object {
                "width": "100%",
              }
            }>
            <td
              className="soft-ends"
              style={
                Object {
                  "fontFamily": "Helvetica",
                }
              }>
              Total
            </td>
            <td />
            <td
              className="soft-ends total-amount"
              style={
                Object {
                  "fontFamily": "Helvetica",
                }
              }>
              $30.00
            </td>
          </tr>
        </tbody>
      </table>
      <p
        className="footer-text soft">
        Thank you for your continued support. Please note that for taxation purposes, no goods or services were received in return for these contributions. If you have any questions or concerns, please contact us at giving@newspring.cc or 864-965-9990.
      </p>
    </div>
  </body>
</html>
`;
